File: Lynx-IOS/Lynx-IOS/Features/Chat/MarkdownMessage.swift
import MarkdownUI
import SwiftUI

/// Renders assistant message markdown with a theme matching the web app's
/// `MarkdownContent` (chat-messages.tsx): mono body text, amber inline code
/// and `›` bullets, dark fenced code blocks, amber blockquote rule.
struct MarkdownMessage: View {
    let content: String
    var streaming: Bool = false

    var body: some View {
        HStack(alignment: .bottom, spacing: 4) {
            Markdown(content)
                .markdownTheme(.lynx)
            if streaming {
                Rectangle()
                    .fill(LynxColor.amberBright)
                    .frame(width: 2, height: 14)
            }
        }
    }
}

private extension Theme {
    static let lynx = Theme()
        .text {
            ForegroundColor(LynxColor.stone200)
            FontFamilyVariant(.monospaced)
            FontSize(14)
        }
        .code {
            FontFamilyVariant(.monospaced)
            FontSize(.em(0.85))
            ForegroundColor(LynxColor.amberBright)
            BackgroundColor(LynxColor.stone800)
        }
        .link {
            ForegroundColor(LynxColor.amberBright)
            UnderlineStyle(.single)
        }
        .strong {
            FontWeight(.bold)
            ForegroundColor(LynxColor.stone100)
        }
        .heading1 { configuration in
            configuration.label
                .markdownTextStyle {
                    FontWeight(.bold)
                    FontSize(17)
                    ForegroundColor(LynxColor.stone100)
                }
                .markdownMargin(top: 16, bottom: 8)
        }
        .heading2 { configuration in
            configuration.label
                .markdownTextStyle {
                    FontWeight(.bold)
                    FontSize(15)
                    ForegroundColor(LynxColor.stone100)
                }
                .markdownMargin(top: 16, bottom: 8)
        }
        .heading3 { configuration in
            configuration.label
                .markdownTextStyle {
                    FontWeight(.semibold)
                    FontSize(14)
                    ForegroundColor(LynxColor.stone200)
                }
                .markdownMargin(top: 12, bottom: 6)
        }
        .paragraph { configuration in
            configuration.label
                .relativeLineSpacing(.em(0.3))
                .markdownMargin(top: 0, bottom: 10)
        }
        .blockquote { configuration in
            configuration.label
                .markdownTextStyle {
                    ForegroundColor(LynxColor.stone400)
                    FontStyle(.italic)
                }
                .padding(.leading, 12)
                .overlay(alignment: .leading) {
                    Rectangle()
                        .fill(LynxColor.amber.opacity(0.6))
                        .frame(width: 2)
                }
                .markdownMargin(top: 8, bottom: 8)
        }
        .bulletedListMarker { _ in
            Text("›")
                .font(LynxFont.mono(13))
                .foregroundStyle(LynxColor.amber)
        }
        .codeBlock { configuration in
            ScrollView(.horizontal, showsIndicators: false) {
                Text(configuration.content)
                    .font(LynxFont.mono(12))
                    .foregroundStyle(LynxColor.stone200)
                    .padding(12)
            }
            .background(LynxColor.codeBg)
            .overlay(
                RoundedRectangle(cornerRadius: 2)
                    .strokeBorder(LynxColor.stone800, lineWidth: 1)
            )
            .clipShape(RoundedRectangle(cornerRadius: 2))
            .markdownMargin(top: 8, bottom: 8)
        }
        .table { configuration in
            configuration.label
                .fixedSize(horizontal: false, vertical: true)
                .markdownMargin(top: 8, bottom: 8)
        }
        .tableCell { configuration in
            configuration.label
                .markdownTextStyle {
                    FontSize(.em(0.85))
                    ForegroundColor(LynxColor.stone300)
                }
                .padding(8)
        }
}

File: Lynx-Andrioid/app/src/main/java/com/anantha/lynx/ui/chat/MarkdownMessage.kt
package com.anantha.lynx.ui.chat

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextLinkStyles
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.anantha.lynx.ui.theme.LynxColor
import com.anantha.lynx.ui.theme.LynxFont
import com.mikepenz.markdown.m3.Markdown
import com.mikepenz.markdown.m3.markdownColor
import com.mikepenz.markdown.m3.markdownTypography

/**
 * Renders assistant message markdown. Ported from
 * Lynx-IOS/Lynx-IOS/Features/Chat/MarkdownMessage.swift.
 *
 * Colors/typography are themed to match Lynx's mono/amber palette (body,
 * headings, code, links, blockquote text). The amber blockquote left-rule
 * and the `›` bullet glyph iOS draws by hand aren't exposed as clean
 * override points in this renderer's public API — left as a known gap
 * rather than guessing at undocumented internals without a device to
 * visually verify against.
 */
@Composable
fun MarkdownMessage(content: String, streaming: Boolean = false) {
    Row(verticalAlignment = Alignment.Bottom) {
        Markdown(
            content,
            colors = markdownColor(
                text = LynxColor.stone200,
                codeBackground = LynxColor.stone800,
                inlineCodeBackground = LynxColor.stone800,
                dividerColor = LynxColor.stone800,
                tableBackground = LynxColor.codeBg,
            ),
            typography = markdownTypography(
                text = LynxFont.mono(14.sp),
                code = LynxFont.mono(12.sp),
                inlineCode = LynxFont.mono(12.sp, FontWeight.Medium).copy(color = LynxColor.amberBright),
                h1 = LynxFont.display(17.sp, FontWeight.Bold).copy(color = LynxColor.stone100),
                h2 = LynxFont.display(15.sp, FontWeight.Bold).copy(color = LynxColor.stone100),
                h3 = LynxFont.display(14.sp, FontWeight.SemiBold).copy(color = LynxColor.stone200),
                h4 = LynxFont.mono(14.sp),
                h5 = LynxFont.mono(13.sp),
                h6 = LynxFont.mono(12.sp),
                quote = LynxFont.mono(14.sp).copy(color = LynxColor.stone400, fontStyle = FontStyle.Italic),
                paragraph = LynxFont.mono(14.sp).copy(color = LynxColor.stone200),
                ordered = LynxFont.mono(14.sp).copy(color = LynxColor.stone200),
                bullet = LynxFont.mono(14.sp).copy(color = LynxColor.stone200),
                list = LynxFont.mono(14.sp).copy(color = LynxColor.stone200),
                textLink = TextLinkStyles(
                    style = SpanStyle(color = LynxColor.amberBright, textDecoration = TextDecoration.Underline),
                ),
                table = LynxFont.mono(12.sp).copy(color = LynxColor.stone300),
            ),
            modifier = Modifier.weight(1f, fill = false),
        )
        if (streaming) {
            Box(modifier = Modifier.width(2.dp).height(14.dp).background(LynxColor.amberBright))
        }
    }
}
